cyberdrop_dl/crawlers/xgroovy.py

```diff
@@ -0,0 +1,133 @@
+            return await self.collection(scrape_item, collection_type)
+        raise ValueError
+
+    @error_handling_wrapper
+    async def gif(self, scrape_item: ScrapeItem) -> None:
+        if await self.check_complete_from_referer(scrape_item):
+            return
+
+        gif_id: str = scrape_item.url.parts[scrape_item.url.parts.index("gifs") + 1]
+        soup = await self.request_soup(scrape_item.url)
+        link = self.parse_url(css.get_attr(css.select_one(soup, _SELECTORS.GIF), "src"))
+        return await self.download_video(scrape_item, gif_id, soup, link)
+
+    @error_handling_wrapper
+    async def video(self, scrape_item: ScrapeItem) -> None:
+        if await self.check_complete_from_referer(scrape_item):
+            return
+
+        video_id: str = scrape_item.url.parts[scrape_item.url.parts.index("videos") + 1]
+        soup = await self.request_soup(scrape_item.url)
+        best_format: Format = _get_best_format(css.select_one(soup, _SELECTORS.VIDEO))
+        return await self.download_video(
+            scrape_item, video_id, soup, self.parse_url(best_format.link_str), resolution=best_format.resolution
+        )
+
+    async def download_video(
+        self,
+        scrape_item: ScrapeItem,
+        file_id: str,
+        soup: BeautifulSoup,
+        link: AbsoluteHttpURL,
+        resolution: str | None = None,
+    ):
+        filename, ext = self.get_filename_and_ext(link.name)
+        title = open_graph.get_title(soup)
+        context = json.loads(css.select_one_get_text(soup, _SELECTORS.UPLOAD_DATE))
+        scrape_item.possible_datetime = self.parse_iso_date(context.get("uploadDate"))
+        custom_filename = self.create_custom_filename(title, ext, file_id=file_id, resolution=resolution)
+        return await self.handle_file(link, scrape_item, filename, ext, custom_filename=custom_filename)
+
+    @error_handling_wrapper
+    async def collection(self, scrape_item: ScrapeItem, collection_type: CollectionType) -> None:
+        title = scrape_item.url.parts[-1]
+        if collection_type == CollectionType.PORNSTARS:
+            soup = await self.request_soup(scrape_item.url)
+            title = css.select_one_get_text(soup, _SELECTORS.PORNSTAR_NAME)
+
+        title = self.create_title(f"{title} - [{collection_type}]")
+        scrape_item.setup_as_album(title)
+        async for soup in self.web_pager(scrape_item.url, _SELECTORS.NEXT_PAGE):
+            for _, new_scrape_item in self.iter_children(scrape_item, soup, _SELECTORS.SEARCH_VIDEOS):
+                self.create_task(self.run(new_scrape_item))
+
+
+def _get_best_format(video_tag):
+    options = []
+    for src in video_tag.find_all("source"):
+        url = src.get("src")
+        title = src.get("title", "0p")
+        resolution = int(title.replace("p", ""))
+        options.append((resolution, url, title))
+    best = max(options, key=lambda x: x[0])
+    return Format(best[2], best[1])
```

Comment:
Make this a generator
```suggestion
   def _get_best_format(video_tag: Tag) -> Format:
    def parse():
        for src in video_tag.select("source"):
            url = get_attr(src, "src")
            title = get_attr_or_none(src, "title") or "0p"
            resolution = int(title.replace("p", ""))
            yield Format(resolution, url)

    return max(parse())

```